node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,734 @@
+	}
+	connBody, err := io.ReadAll(connResp.Body)
+	if err != nil {
+		return connIdMap, fmt.Errorf("read failed: %w", err)
+	}
+	connResp.Body.Close()
+
+	var result ibcAllChainConnectionsQueryResults
+	err = json.Unmarshal(connBody, &result)
+	if err != nil {
+		return connIdMap, fmt.Errorf("failed to unmarshal response: %s, error: %w", string(connBody), err)
+	}
+
+	if len(result.Data.ChainConnections) == 0 {
+		return connIdMap, fmt.Errorf("query did not return any data")
+	}
+
+	for idx, connData := range result.Data.ChainConnections {
+		if len(connData) != 2 {
+			return connIdMap, fmt.Errorf("connection map entry %d contains %d items when it should contain exactly two, json: %s", idx, len(connData), string(connBody))
+		}
+
+		connectionIdBytes, err := base64.StdEncoding.DecodeString(connData[0].(string))
+		if err != nil {
+			return connIdMap, fmt.Errorf("connection ID for entry %d is invalid base64: %s, err: %s", idx, connData[0], err)
+		}
+
+		connectionID := string(connectionIdBytes)
+		chainID := vaa.ChainID(connData[1].(float64))
+		connIdMap[chainID] = connectionID
+		w.logger.Debug("IBC connection ID mapping", zap.String("connectionID", connectionID), zap.Uint16("chainID", uint16(chainID)))
+	}
+
+	return connIdMap, nil
+}
+
+// getConnectionID returns the IBC connection ID associated with the given IBC channel. It uses a cache to avoid constantly
+// querying worm chain. This works because once an IBC channel is closed its ID will never be reused.
+func (w *Watcher) getConnectionID(channelID string) (string, error) {
+	w.connectionIdLock.Lock()
+	defer w.connectionIdLock.Unlock()
+	connectionID, exists := w.connectionIdMap[channelID]
+	if exists {
+		return connectionID, nil
+	}
+
+	connectionID, err := w.queryConnectionID(channelID)
+	if err != nil {
+		return connectionID, err
+	}
+
+	w.connectionIdMap[channelID] = connectionID
+	return connectionID, nil
+}
+
+/*
+This query:
+  http://localhost:1319/ibc/core/channel/v1/channels/channel-0/ports/wasm.wormhole1nc5tatafv6eyq7llkr2gv50ff9e22mnf70qgjlv737ktmt4eswrq0kdhcj
+
+Returns something like this:
+{
+  "channel": {
+    "state": "STATE_OPEN",
+    "ordering": "ORDER_UNORDERED",
+    "counterparty": {
+      "port_id": "wasm.terra14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9ssrc8au",
+      "channel_id": "channel-0"
+    },
+    "connection_hops": [
+      "connection-0"
+    ],
+    "version": "ibc-wormhole-v1"
+  },
+  "proof": null,
+  "proof_height": {
+    "revision_number": "0",
+    "revision_height": "358"
+  }
+}
+*/
+
+// ibcChannelQueryResults is used to parse the result from the IBC connection ID query.
+type ibcChannelQueryResults struct {
+	Channel struct {
+		State          string
+		ConnectionHops []string `json:"connection_hops"`
+		Version        string
+	}
+}
+
+// getConnectionID queries the contract on wormchain to map a channel ID to a connection ID.
+func (w *Watcher) queryConnectionID(channelID string) (string, error) {
+	client := &http.Client{
```

Comment:
This causes a request per message right now. Any reason you didn't opt to cache this.

Reply:
The caching is done one level up in `getConnectionID`. This function is only called when we actually need to do the query.